### PR TITLE
Test the func args invariant in test_args

### LINKED ISSUE #8210: Zero degree polynomial copy() error
I have a list of polynomials that I need to `deepcopy()`.  But thet procedure generates error on polynomials of zero degree. For example:

```
In [1]: p = Poly([0], x)
In [2]: p.copy() 
```

I get  `GeneratorsNeeded: specify generators to give 0 a meaning`. And I want to have that `Poly([0],x)` as a neutral element of addition.

Temporarily I use something like `Poly([10**(-10), 0], x)` but it is not very elegant, is it?

### LINKED ISSUE #18529: DimensionSystem arg handling inconsistency
@Upabjojr there is a problem in `DimensionSystem.__new__` that means it does not satisfy the `obj == obj.func(*obj.args)` invariant. Demo:
```python
from sympy.physics.units.dimensions import DimensionSystem
from sympy.physics.units.definitions.dimension_definitions import length, time, velocity
ds = DimensionSystem((length, time), (velocity,))
print(ds)
print(ds.func(*ds.args))
```
Output:
```
DimensionSystem((Dimension(length, L), Dimension(time, T)),
    (Dimension(velocity),), {length: {length: 1}, time: {time: 1}, Dimension(velocity): {velocity: 1}})
DimensionSystem((Dimension(length, L), Dimension(time, T)),
    (Dimension(velocity),), {length: {length: 1}, time: {time: 1}, velocity: {velocity: 1}})
```
(Note velocity rather than Dimension(velocity).)

_Originally posted by @oscarbenjamin in https://github.com/sympy/sympy/pull/18513#issuecomment-580946465_

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### References to other Issues or PRs

Closes #10128
Fixes #18529 
Fixes #8210 

#### Brief description of what is fixed or changed

Tests the `obj = obj.func(*obj.args)` invariant in `test_args.py`. Fixes args for some classes.

NO ENTRY

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
@Upabjojr there is a problem in `DimensionSystem.__new__` that means it does not satisfy the `obj == obj.func(*obj.args)` invariant. Demo:
```python
from sympy.physics.units.dimensions import DimensionSystem
from sympy.physics.units.definitions.dimension_definitions import length, time, velocity
ds = DimensionSystem((length, time), (velocity,))
print(ds)
print(ds.func(*ds.args))
```
Output:
```
DimensionSystem((Dimension(length, L), Dimension(time, T)), (Dimension(velocity),), {length: {length: 1}, time: {time: 1}, Dimension(velocity): {velocity: 1}})
DimensionSystem((Dimension(length, L), Dimension(time, T)), (Dimension(velocity),), {length: {length: 1}, time: {time: 1}, velocity: {velocity: 1}})
```
(Note velocity rather than Dimension(velocity).)

**Comment by Upabjojr** [comment]
OK. Can you open an issue and assign it to me? As soon as I have some time I'll try to fix it.

**Comment by oscarbenjamin** [comment]
Okay this is ready for review.

Among other things this fixes `Poly.args` so that it contains the generators as well as the expression e.g. on master:
```julia
In [1]: p = Poly(x, y)                                                                                                                         

In [2]: p.args                                                                                                                                 
Out[2]: (x,)

In [3]: p.free_symbols                                                                                                                         
Out[3]: {x}

In [4]: p.has(y)                                                                                                                               
Out[4]: False

In [5]: Poly(0, x).copy()                                                                                                                         
---------------------------------------------------------------------------
GeneratorsNeeded                          Traceback (most recent call last)
```
With the PR we have
```julia
In [1]: p = Poly(x, y)                                                                                                                         

In [2]: p.args                                                                                                                                 
Out[2]: (x, y)

In [3]: p.free_symbols                                                                                                                         
Out[3]: {x}

In [4]: p.has(y)                                                                                                                               
Out[4]: True

In [5]: Poly(0, x).copy()                                                                                                                      
Out[5]: Poly(0, x, domain='ZZ')
```
I'm not sure whether the generators should be part of `free_symbols` or not. We have
```julia
In [6]: Poly(0, x).subs(x, y)                                                                                                                  
Out[6]: Poly(0, y, domain='ZZ')
```
which suggests that it should be. Now that `Poly.args` is fixed I can delete the `Poly.free_symbols` method to get
```julia
In [2]: Poly(x, y).free_symbols                                                                                                                
Out[2]: {x, y}
```

Note that this PR adds both the polynomial expression (Expr instance) and generators to the args passed to `Basic.__new__`. That means that these are explicitly stored as a straight-forward args tuple for all Poly instances. It seems deliberate that `p = Poly(expr, x)` does not actually put `expr` in `p._args[0]` presumably for performance reasons. I see no measurable slowdown as a result though.

**Comment by oscarbenjamin** [comment]
This needs a review

**Comment by jksuom** [comment]
I'm not sure that `expr` and `gens` will suffice to recreate a `Poly` object in general; keyword arguments, or at least some of them, are probably also needed. Some attributes like `domain` can only be determined from some keyword parameters. They are currently stored in the `rep`attribute but it is not clear how that could be included among the parameters.

**Comment by oscarbenjamin** [comment]
It isn't possible to include non-Basic domain in the args but it would be if domains had a sympified representation. I left that part out because it can be done but is a bigger task. Provided the domain was automatically chosen from the expr and gens arguments then this PR will ensure that a Poly can be rebuilt from its args. Otherwise the domain may not be the same after rebuilding.

**Comment by oscarbenjamin** [comment]
We can do:
```
In [4]: Poly(x*y, (x, y))                                                                                                                      
Out[4]: Poly(x*y, x, y, domain='ZZ')
```
So maybe the args tuple for `Poly` should be `(expr, gens)` with a tuple of generators as the second argument. Then it would be possible to extend it to a 3-tuple with the domain if the domain had a sympified Basic representation.

**Comment by oscarbenjamin** [comment]
Currently using 3-args with a tuple as second arg raises so that could be changed to accept the domain as a position argument:
```
In [3]: Poly(x, (x,), 'ZZ')                                                                                                                  
---------------------------------------------------------------------------
AttributeError 
```
I wonder what the desired behaviour is though when e.g. using subs with a Poly. For example
```
In [14]: Poly(x*y, x, domain='EX').xreplace({y:1})                                                                                             
Out[14]: Poly(x, x, domain='ZZ')
```
Here the domain has been recalculated by xreplace. If the domain was a proper part of args then it would be propagated by xreplace.

**Comment by oscarbenjamin** [comment]
The other weird thing that `Poly` does is convert to a plain expression when subs is called:
```
In [10]: p = Poly(x*y, (x,))                                                                                                                   

In [11]: p                                                                                                                                     
Out[11]: Poly(y*x, x, domain='ZZ[y]')

In [12]: p.subs(y, 1)                                                                                                                          
Out[12]: x
```
I think that should ultimately be fixed but it would not be a backward compatible change.

**Comment by oscarbenjamin** [comment]
@jksuom I want this change (testing the func args invariant) in. I'm happy to revert or alter the changes to `Poly` if needed.

The change to Poly here is just that args also includes the generators. I think a better approach is to have the generators as a tuple in the args though so e.g. this PR currently gives:
```
In [1]: Poly(x*y).args                                                                                                                         
Out[1]: (x⋅y, x, y)
```
But I think it would be a little better if the Poly args were like this
```
In [3]: Basic(x*y, (x, y)).args                                                                                                                
Out[3]: (x⋅y, (x, y))
```

It is also possibly desirable to include the domain in args but that isn't possible currently and is a bigger job. We need to define how we want the domain to be defined after a substitution. There are two obvious options:

1. The domain should be unchanged after any operation like subs/xreplace etc. Then we would have
```
In [4]: subs = {y: 1}                                                                                                                          

In [5]: Poly(x*y, (x,)).domain                                                                                                                 
Out[5]: ZZ[y]

In [7]: Poly(x*y, (x,)).xreplace(subs).domain # On master this gives ZZ
Out[7]: ZZ[y]
```

2. The domain should be recalculated from the args each time the Poly is rebuilt. That is what xreplace currently does on master and meets the invariant:
```
In [9]: Poly((x*y).xreplace(subs), (x,)).domain == Poly(x*y, (x,)).xreplace(subs).domain                                                       
Out[9]: True
```

Going with option 2 implies that we don't want the domain to be part of the args since it should be recalculated from the other args after any operation. Then a Poly with an explicitly specified alternate domain becomes an exception to the func-args invariant much like instances created with evaluate=False e.g. with this PR:
```
In [11]: p = Poly(x, (x,), domain='QQ')                                                                                                        

In [12]: p                                                                                                                                     
Out[12]: Poly(x, x, domain='QQ')

In [13]: p2 = p.func(*p.args)                                                                                                                  

In [14]: p2                                                                                                                                    
Out[14]: Poly(x, x, domain='ZZ')

In [15]: p2 == p2.func(*p2.args)                                                                                                               
Out[15]: True
```
This is analogous to
```
In [16]: m = Mul(2, 3, evaluate=False)                                                                                                         

In [17]: m                                                                                                                                     
Out[17]: 2⋅3

In [18]: m.func(*m.args)                                                                                                                       
Out[18]: 6
```

The current arrangement for Poly breaks the hash-eq invariant:
```
In [2]: p1 = Poly(x, x, domain='ZZ')                                                                                                           

In [3]: p2 = Poly(x, x, domain='QQ')                                                                                                           

In [4]: p1                                                                                                                                     
Out[4]: Poly(x, x, domain='ZZ')

In [5]: p2                                                                                                                                     
Out[5]: Poly(x, x, domain='QQ')

In [6]: p1 == p2                                                                                                                               
Out[6]: True

In [10]: hash(p1) == hash(p2)                                                                                                                  
Out[10]: False
```
If that were fixed then Polys with different domains would become interchangeable under caching. If we want the domain to be a meaningful part of the object then Polys with different domains should compare unequal. That to me suggests that we need to go with option 1. above.

**Comment by oscarbenjamin** [comment]
In summary I think that:

1. There should be Basic classes corresponding to the domain types so that domains are sympifiable
2. Poly.args should be (expr, gens, domain)
3. Poly.args should be stored in `_args` like normal `Basic` types.
4. Poly instances with different domains should compare unequal with `==`
5. Poly should not redefine Basic methods like args, subs, `__eq__`, `__hash__`.
6. PurePoly should do nothing more than canonicalise the symbols

However I don't propose to do all those things in this PR. Here I mainly want to ensure that we can test that new classes are not breaking the args invariant.

**Comment by jksuom** [comment]
Those suggestions look good to me, and I believe that they can be implemented without too much trouble. The hardest part may be `domain`.

> I want this change (testing the func args invariant) in. 

These changes look ok to me. It should not be necessary to have everything in one PR.

**Comment by oscarbenjamin** [comment]
Thanks!

File: sympy/polys/polytools.py [hunk 1]
```diff
@@ -6838,7 +6826,8 @@ def _new(cls, basis, options):
 
     @property
     def args(self):
-        return (Tuple(*self._basis), Tuple(*self._options.gens))
+        basis = (p.as_expr() for p in self._basis)
```
Comment (asmeurer) [inline sympy/polys/polytools.py:6829]: @oscarbenjamin I'm guessing this is where the performance regression issues come from, though I didn't verify it. Would it help to cache this?

Comment (oscarbenjamin) [reply sympy/polys/polytools.py:6829]: I wonder if we should use `cached_property` generally now that it's in the stdlib:
https://docs.python.org/dev/library/functools.html?highlight=s#functools.cached_property

File: sympy/polys/polytools.py [hunk 2]
```diff
@@ -133,13 +133,21 @@ def new(cls, rep, *gens):
         elif rep.lev != len(gens) - 1:
             raise PolynomialError("invalid arguments: %s, %s" % (rep, gens))
 
-        obj = Basic.__new__(cls)
+        expr = basic_from_dict(rep.to_sympy_dict(), *gens)
 
+        obj = Basic.__new__(cls, expr, *gens)
```
Comment (oscarbenjamin) [inline sympy/polys/polytools.py:138]: Calling `basic_from_dict` here would be my guess at a regression

**Pull request merged**